src/System.IO.Pipelines.Networking.Tls/ApplicationProtocols.cs

```diff
@@ -0,0 +1,157 @@
+﻿using System.IO.Pipelines.Networking.Tls.Internal.Sspi;
+
+namespace System.IO.Pipelines.Networking.Tls
+{
+    public static class ApplicationProtocols
+    {
+        private static readonly byte[] _http1 = {0x68, 0x74, 0x74, 0x70, 0x2f, 0x31, 0x2e, 0x31}; //("http/1.1")
+        private static readonly byte[] _spdy1 = {0x73, 0x70, 0x64, 0x79, 0x2f, 0x31}; //("spdy/1")
+        private static readonly byte[] _spdy2 = {0x73, 0x70, 0x64, 0x79, 0x2f, 0x32}; //("spdy/2")
+        private static readonly byte[] _spdy3 = {0x73, 0x70, 0x64, 0x79, 0x2f, 0x33}; //("spdy/3")
+
+        private static readonly byte[] _traversalUsingRelaysAroundNAT =
+        {
+            0x73, 0x74, 0x75, 0x6E, 0x2E, 0x74, 0x75, 0x72,
+            0x6E
+        }; //("stun.turn")
+
+        private static readonly byte[] _natDiscoveryUsingSessionTraversalUtilitiesforNAT =
+        {
+            0x73, 0x74, 0x75, 0x6E, 0x2E,
+            0x6e, 0x61, 0x74, 0x2d, 0x64, 0x69, 0x73, 0x63, 0x6f, 0x76, 0x65, 0x72, 0x79
+        }; //("stun.nat-discovery")
+
+        private static readonly byte[] _http2overTLS = {0x68, 0x32}; //("h2")
+        private static readonly byte[] _http2overTCP = {0x68, 0x32, 0x63}; //("h2c")
+        private static readonly byte[] _webRTCMediaAndData = {0x77, 0x65, 0x62, 0x72, 0x74, 0x63}; //("webrtc")
+
+        private static readonly byte[] _confidentialWebRTCMediaAndData =
+        {
+            0x63, 0x2d, 0x77, 0x65, 0x62, 0x72, 0x74, 0x63
+        }; //("c-webrtc")
+
+        private static readonly byte[] _Ftp = {0x66, 0x74, 0x70}; //("ftp")
+
+        private static readonly byte[][] _allProtocols =
+        {
+            _http1, _spdy1, _spdy2, _spdy3, _traversalUsingRelaysAroundNAT,
+            _natDiscoveryUsingSessionTraversalUtilitiesforNAT, _http2overTLS, _http2overTCP, _webRTCMediaAndData,
+            _confidentialWebRTCMediaAndData, _Ftp
+        };
+
+        private static readonly Array _numberOfProtocols = Enum.GetValues(typeof(ProtocolIds));
+
+        const int _HeaderLength = 10;
+        const int _ContentOffset = 4;
+
+        [Flags]
+        public enum ProtocolIds
+        {
+            None = 0,
+            Http11 = 1,
+            Spdy1 = 2,
+            Spdy2 = 4,
+            Spdy3 = 8,
+            TraversalUsingRelaysAroundNat = 16,
+            NatDiscoveryUsingSessionTraversalUtilitiesforNat = 32,
+            Http2OverTls = 64,
+            Http2OverTcp = 128,
+            WebRtcMediaAndData = 256,
+            ConfidentialWebRtcMediaAndData = 512,
+            Ftp = 1024
+        }
+
+        internal enum ApplicaitonProtocolNegotiationExtension
+        {
+            None = 0,
+            Npn,
+            Alpn,
+        }
+
+        internal enum ApplicationProtocolNegotiationStatus : uint
+        {
+            None = 0,
+            Success,
+            SelectedClientOnly
+        }
+
+        internal static unsafe ProtocolIds FindNegotiatedProtocol(SSPIHandle context)
```

Comment:
there is a pointer in the SSPIHandle to the buffer with the byte array in it. Now gets pushed into the span before it calls the underlying method.